AGENTS.md

```diff
@@ -0,0 +1,1422 @@
+  describe('addToken', () => {
+    it('adds the token to state', () => {
+      // Arrange, Act, Assert
+    });
+
+    it('throws error when token address is missing', () => {
+      // Test error case
+    });
+  });
+});
+```
+
+**Detailed Guidelines:** See `.cursor/rules/unit-testing-guidelines.mdc`
+
+### E2E Tests
+
+**Location:** `test/e2e/tests/`
+
+**Running:**
+```bash
+# Must build test build first!
+yarn build:test              # or yarn start:test
+
+# Run E2E tests
+yarn test:e2e:chrome         # All Chrome tests
+yarn test:e2e:firefox        # All Firefox tests
+
+# Single test with debug
+yarn test:e2e:single test/e2e/tests/TEST_NAME.spec.js \
+  --browser=chrome \
+  --debug \
+  --leave-running
+```
+
+**Options:**
+- `--browser` - chrome, firefox, or all
+- `--debug` - Verbose logging
+- `--leave-running` - Keep browser open on failure
+- `--retries` - Number of retries on failure
+- `--update-snapshot` - Update snapshots
+
+**E2E Best Practices:**
+- Always use test builds (not dev builds)
+- Tests should be independent and isolated
+- Use page objects for reusable UI interactions
+- Clean up state between tests
+
+### Integration Tests
+
+**Location:** `test/integration/`
+
+**Running:**
+```bash
+yarn test:integration
+yarn test:integration:coverage
+```
+
+**Coverage Goals:**
+- Unit tests: > 80% coverage
+- Critical paths: > 90% coverage
+- E2E tests: Cover main user workflows
+
+---
+
+## State Migrations
+
+### What are Migrations?
+
+When MetaMask updates, the stored state format might change. Migrations transform old state to new format automatically.
+
+### Creating a Migration
+
+```bash
+# Generate migration template
+yarn generate:migration
+
+# Creates: app/scripts/migrations/XXX.ts (next number)
+```
+
+### Migration Guidelines
+
+1. **Always create migrations for state changes**
+2. **Test migrations thoroughly** (old state → new state)
+3. **Handle missing data gracefully** (some users may have old/corrupted state)
+4. **Never mutate input state** (return new state object)
+5. **Include version number** in migration metadata
+
+**Example Migration:**
+```typescript
+import { cloneDeep } from 'lodash';
+
+const version = 123;
+
+export default {
+  version,
+  async migrate(originalVersionedData: any) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    transformData(versionedData.data);
+    return versionedData;
+  },
+};
+
+function transformData(state: any): void {
+  // Transform state.data
+  if (state.PreferencesController) {
+    state.PreferencesController.newProperty = 'defaultValue';
+  }
+}
+```
+
+---
+
+## Pull Request Workflow
+
+### Before Creating a PR
+
+- [ ] All tests pass: `yarn test`
+- [ ] Linting passes: `yarn lint`
+- [ ] No console.logs or debug code
+- [ ] Changes are covered by tests
+- [ ] LavaMoat policies updated (if dependencies changed)
+- [ ] Attributions updated (if dependencies changed)
+
+### Creating a PR
```

Comment:
Here's I'd say the AI should create a PR description that exactly follows the [PR template](https://github.com/MetaMask/metamask-extension/blob/main/.github/pull-request-template.md).